### ISSUE TITLE
Make NodeAffinity mutable for Jobs

### ISSUE THREAD

**Issue by ahg-g** [kubernetes/kubernetes#104714]
#### What would you like to be added:

Make NodeAffinity constraints mutable for suspended Jobs. 

#### Why is this needed:

In https://github.com/kubernetes/kubernetes/issues/97170 we added a new flag to allow suspending jobs to control when the Pods of a Job get created by controller-manager. We proposed this as a primitive that higher-level controllers can use to support job queuing,

Allowing to mutate NodeAffinity constraints allows a higher-level queue controller to influence job placement by injecting node affinity constraints before un-suspending a job.

The implementation is basically relaxing the validation on job updates to allow for mutating the node affinity constraints of the pod template for suspended jobs.

/sig scheduling
/sig apps
/kind feature

**Comment by Huang-Wei** [comment]
Two questions:

1. By "nodeAffinity", I suppose it means `pod.spec.affinity.nodeAffinity` - with volume affinity excluded. Given it's a pod field, this would allow arbitrary workload (not only Job) to mutate it, is it intentional, or you want to restrict to Job API only?
2. You mentioned it's intended for suspended Job, so are you going to limit the scope to pending Pods (with empty `.spec.nodeName`) only? - mutating (on purpose or by mistake) a running Pod's nodeAffinity may cause unexpected behavior.

**Comment by ahg-g** [comment]
> Two questions:
> 
> 1. By "nodeAffinity", I suppose it means `pod.spec.affinity.nodeAffinity` - with volume affinity excluded. Given it's a pod field, this would allow arbitrary workload (not only Job) to mutate it, is it intentional, or you want to restrict to Job API only?

Yeah, I am referring to nodeAffinity constraints in the pod template of the job. It is restricted to the job api only at this point.

> 2. You mentioned it's intended for suspended Job, so are you going to limit the scope to pending Pods (with empty `.spec.nodeName`) only? - mutating (on purpose or by mistake) a running Pod's nodeAffinity may cause unexpected behavior.

The proposal is not to allow mutating nodeAffinity of pods, only the pod template of a suspended job. So when the job is unsuspended, the pods that will be created will already be carrying the modified constraints.

**Comment by Huang-Wei** [comment]
Thanks for the clarification. The limited scope is fine with me. I will leave technical feasibility and to API experts.

> So when the job is unsuspended, the pods that will be created will already be carrying the modified constraints.

I suppose this is based on the current semantics of "suspended job" - when a Job is suspended, it deletes all pods it always spawns. So what if the semantics change in the future to "literally only spawn new pods and keep created pods intact"?

**Comment by ahg-g** [comment]
>  So what if the semantics change in the future to "literally only spawn new pods and keep created pods intact"?

I don't think this is an acceptable change of existing behavior of the `suspended` feature; if that to happen, it will need to be done using a new API field.

**Comment by liggitt** [comment]
Letting the scheduling portions of the pod template (node selector, node affinity, anti-affinity, tolerations) be mutable until the job is unsuspended could make sense for late-binding scheduling purposes.

Could the mutation be limited only be allowed before *the* job is unsuspended? Ideally we could avoid this scenario:

1. create an unsuspended job
2. controller starts creating pods
3. suspend the job
4. change the pod template
5. unsuspend the job
6. controller handling does not fully delete pods created in step 2 (controller was down, or events arrived quickly and job controller sync handler was busy with other jobs)
7. controller continues creating pods from new template

The result of that is that you have a mix of pods for the same job with different specs/scheduling directives

**Comment by ahg-g** [comment]
Perhaps we can avoid the scenario by accepting the mutation only if the the job is suspended and the number of [active pods](https://github.com/kubernetes/kubernetes/blob/7997805f339923451d069ba12470bbc79059cf36/pkg/apis/batch/types.go#L248) is 0?

**Comment by liggitt** [comment]
> Perhaps we can avoid that scenario by accepting the mutation only if the the job is suspended and the number of [active pods](https://github.com/kubernetes/kubernetes/blob/7997805f339923451d069ba12470bbc79059cf36/pkg/apis/batch/types.go#L248) is 0?

does the controller set active pods *before* creating new pods and after deleting pods for a suspended job? if not, there's a race

**Comment by ahg-g** [comment]
> > Perhaps we can avoid that scenario by accepting the mutation only if the the job is suspended and the number of [active pods](https://github.com/kubernetes/kubernetes/blob/7997805f339923451d069ba12470bbc79059cf36/pkg/apis/batch/types.go#L248) is 0?
> 
> does the controller set active pods _before_ creating new pods and after deleting pods for a suspended job? if not, there's a race

I believe so, @alculquicondor can you confirm?

**Comment by ahg-g** [comment]
Actually it seems that in both creating and deleting the status is updated after. But I think the create part isn't really related to our case, we only care about the delete case.

**Comment by alculquicondor** [comment]
Yes, first the pods are created/deleted, and then the status field is updated if there were no errors.

@liggitt why do we need to restrict the change to suspended jobs? If the user changes the node affinity in a way that the job stops working, that's user error. But generally, the job would continue working. The pods just might start going to different zones.

In fact, I can imagine a use case beyond suspended jobs:

    I have a long running job (high number of completions and relatively low parallelism) that ended up with affinity to certain zone. Then, they announce that the zone is going to go on maintenance. I want the new pods to start in a different zone, but I want to let the old pods finish successfully. I didn't suspend the job at any point.

**Comment by Huang-Wei** [comment]
@alculquicondor if you don't restrict it to suspended jobs, in your case (updating an **unsuspended** job's template), isn't the running pods' spec gets updated and then kubelet may reject them? (by kubelet's regular admission check)

**Comment by ahg-g** [comment]
No, pods that have already been created isn't impacted by the change.

I don't have a strong opinion whether or not we restrict the change to suspended jobs.

**Comment by wojtek-t** [comment]
> Actually it seems that in both creating and deleting the status is updated after. But I think the create part isn't really related to our case, we only care about the delete case.

It is related, imagine:
- we unsuspend the job
- we create the pod (but they job is not yet updated)
- we suspend again
- we change the node-affinity

And we have pods with different affinities.

I think that if we can prevent the race, I would be fine with it, but if not I agree with what Jordan suggested.

**Comment by ahg-g** [comment]
> > Actually it seems that in both creating and deleting the status is updated after. But I think the create part isn't really related to our case, we only care about the delete case.
> 
> It is related, imagine:
> 
> * we unsuspend the job
> * we create the pod (but they job is not yet updated)
> * we suspend again
> * we change the node-affinity
> 
> And we have pods with different affinities.
> 

Can you elaborate? The condition is suspended + zero active pods, and so changing affinity will only be allowed after all existing pods are deleted (active = 0)

**Comment by ahg-g** [comment]
I updated the issue to include tolerations. I feel that both node affinity and tolerations are node related scheduling directives, while pod affinity and topology spread are more workload specific that ideally should be declared by the workload itself.

**Comment by liggitt** [comment]
> @liggitt why do we need to restrict the change to suspended jobs? If the user changes the node affinity in a way that the job stops working, that's user error. But generally, the job would continue working. The pods just might start going to different zones.

I guess it depends on the user's expectations... if we let them change scheduling bits on the pod template, will it surprise them if there are pending unschedulable pods for the job with the old scheduling bits?

> I updated the issue to include tolerations. I feel that both node affinity and tolerations are node related scheduling directives, while pod affinity and topology spread are more workload specific that ideally should be declared by the workload itself.

Makes sense to me. I'd also include nodeSelector in the set of node-related scheduling directives that should be mutable.

**Comment by alculquicondor** [comment]
You can't prevent the race condition. Elaborating on @wojtek-t's scenario:

- we unsuspend the job
- the job controller creates the pod, but the job status update fails (so active remains at zero).
- we suspend again
- we change the node-affinity (note that this needs to happen in apiserver, were we wouldn't be listing pods during the update validation)

I think it's better to not restrict the change to suspended jobs and be clear that existing pods are not re-created.

**Comment by wojtek-t** [comment]
> Can you elaborate? The condition is suspended + zero active pods, and so changing affinity will only be allowed after all existing pods are deleted (active = 0)

Aldo pointed one race. But even without failing job status there is a race.

Creation of a pod and Job status update aren't synchronous, and creation of a pod happens first.

So you can in theory always put:
"suspend job + change node selector + unsuspend"
API calls in the middle.

**Comment by wojtek-t** [comment]
> I think it's better to not restrict the change to suspended jobs and be clear that existing pods are not re-created.

It's a cleaner semantic, but it might be surprising/confusing to users, I'm afraid.

**Comment by ahg-g** [comment]
I guess the only way to ensure no surprises is if we allow mutating affinity only if the job started as suspended and was never unsuspended when making the update, I don't think we have something in the job status indicating that though. But that also might be too restrictive.

@wojtek-t @liggitt are we ok then with suspended+active=0 condition while acknowledging the potential race? I am also on the fence with regards to allowing it with no restrictions, but perhaps it should be fine with proper documentation emphasizing best practices on how/when to update those directives?

**Comment by wojtek-t** [comment]
I personally prefer "no restrictions", than a racy one of "suspended + active=0"

**Comment by liggitt** [comment]
> are we ok then with suspended+active=0 condition

I agree. I don't think we should make the spec API consumer poll for status before allowing them to set spec fields... if we want to allow scheduling fields to be mutable throughout the job lifetime, I think the controller should be made to deal with that.

The clearest scenario is the `suspended job → targeted job → unsuspended job` progression. I don't have a good sense for how common retargeting an already running job is to know if it's worth the extra work to make the job controller handle that reliably.

**Comment by ahg-g** [comment]
> The clearest scenario is the suspended job → targeted job → unsuspended job progression

Right, and that would be the best practice we could highlight in the docs.

Sounds good then, I will update the KEP to remove the restriction and add node selector.

**Comment by ahg-g** [comment]
hmm, @liggitt I misread your comment, it seems you still prefer we check for suspended+active=0, correct?

**Comment by liggitt** [comment]
> it seems you still prefer we check for suspended+active=0 during validation, correct?

not quite... I think we should either:
1. only allow mutation for jobs which have never been unsuspended
2. allow mutation in all cases and make the job controller handle mutated scheduling fields in the template in the least surprising way possible, and make the guaranteed behavior super clear in the API docs (*I'd* be surprised if I changed scheduling fields in the template and my job got stuck with one unschedulable pending pod with the old scheduling values)

> Note that the job controller is not involved here, it should simply react to the new state we allowed into the api-server.

If we make active==0 part of the required state to be allowed to mutate these fields, it is involved (you'd have to suspend, wait for job controller to observe, delete, update status to active=0 before being allowed to mutate)

**Comment by ahg-g** [comment]
> only allow mutation for jobs which have never been unsuspended

This feels too restrictive. I still want to support the case of preempting jobs and send them somewhere else.

> allow mutation in all cases and make the job controller handle mutated scheduling fields in the template in the least surprising way possible, and make the guaranteed behavior super clear in the API docs (I'd be surprised if I changed scheduling fields in the template and my job got stuck with one unschedulable pending pod with the old scheduling values)

The least surprising way with guaranteed behavior is one of two options: 
1) the job-controller re-creates all pods, so basically a change to the template is implicitly a suspend. If we do this, then we are not enabling the re-targeting use case while allowing existing pods to complete. 
2) a slight enhancement of the one above, we only re-create pods that are not scheduled yet.

> If we make active==0 part of the required state to be allowed to mutate these fields, it is involved (you'd have to suspend, wait for job controller to observe, delete, update status to active=0 before being allowed to mutate)

ok, to achieve this the job-controller will need to mark the job as "reconciling" whenever the job.status is not yet matched with the actual status of the job or update the status in general before creating pods, which is not going to be ideal in cases where in the same cycle the controller is deleting some pods and adding others (not common, but possible).


I am torn here, it feels like an overkill to be honest. But perhaps re-creating the pods (all or only unscheduled ones) might be the best option to move forward here if we the option of simply documenting the behavior is ruled out.

**Comment by soltysh** [comment]
> The clearest scenario is the `suspended job → targeted job → unsuspended job` progression. I don't have a good sense for how common retargeting an already running job is to know if it's worth the extra work to make the job controller handle that reliably.

When I was initially thinking about it I didn't consider the race (pods with different scheduling directives) being a problem, although now I question my assumption. I guess it will depend on the particular use-case, the job's pod will do whatever it's told no matter where it lands, or in the worst case it'll fail and will be re-created (if specialized hardware was required and no such scheduling directive was initially presented, for example). 
I'm curious to learn how many such situations we're talking about before we invest effort into dealing with that race, which will unnecessarily complicate the controller. Is it worth the effort? 
Personally, I'm leaning towards best effort approach which @ahg-g proposed, ie. suspended + active=0 as being sufficient, assuming it will be properly documented.

**Comment by ahg-g** [comment]
> I'm curious to learn how many such situations we're talking about before we invest effort into dealing with that race, which will unnecessarily complicate the controller. Is it worth the effort?

At scale such races surely will manifest itself. My worry is that to prevent this race condition and guarantee that the job is suspended we will not only need to complicate the job controller, but also very likely incur at least one additional API call per job, which I would very much like to avoid for such a feature that may not be used frequently.

I expect the most common use case is indeed to update affinity before a job has ever been unsuspended; note that a race would still exist even if we try implement such a constraint because it also involves checking the job status's StartTime, and so the same race conditions discussed above apply. However, I think this is a race that is less likely to occur unintentionally. 

So I suggest we do this in two phases:

Phase 1: allow the update for suspended jobs with StartTime=nil (meaning the job is never started before), acknowledging the potential race if someone intentionally try to trigger it.

Phase 2 (if there is demand): allow the update with no constraints while offering guarantees on behavior via re-creating pods (potentially re-creating only the not yet scheduled ones). This is backward compatible with phase 1.

The reason I don't want to purse phase 2 now is that the job controller has become complicated and it is due for a refactor, which will naturally happen when we graduate https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2307-job-tracking-without-lingering-pods to GA.

What do you folks think?

**Comment by wojtek-t** [comment]
> Phase 1: allow the update for suspended jobs with StartTime=nil (meaning the job was never started before), acknowledging the potential race if someone intentionally try to trigger it.

I like this - let's do that and figure out later if we really need more.

**Comment by soltysh** [comment]
:+1: I totally agree with simpler approach and I'm similarly worried about additional complication of the controller. Thank you very much @ahg-g!

**Comment by xiongzubiao** [comment]
Should the `schedulerName` be allowed to change for suspended Jobs as well? Technically `schedulerName` is also a part of the scheduling directives... It can be useful when a different scheduler must be used on a subset of nodes.
